src/Transports.AspNetCore/WebSockets/WebSocketConnection.cs

```diff
@@ -0,0 +1,253 @@
+#pragma warning disable CA2208 // Instantiate argument exceptions correctly
+                throw new ArgumentOutOfRangeException(nameof(options) + "." + nameof(GraphQLWebSocketOptions.DisconnectionTimeout));
+#pragma warning restore CA2208 // Instantiate argument exceptions correctly
+        }
+        _closeTimeoutMs = (int)(options.WebSockets.DisconnectionTimeout ?? DefaultDisconnectionTimeout).TotalMilliseconds;
+        _webSocket = webSocket ?? throw new ArgumentNullException(nameof(webSocket));
+        _stream = new(webSocket);
+        _serializer = serializer ?? throw new ArgumentNullException(nameof(serializer));
+        _pump = new(HandleMessageAsync);
+        RequestAborted = cancellationToken;
+    }
+
+    /// <summary>
+    /// Listens to incoming messages on the WebSocket specified in the constructor,
+    /// dispatching the messages to the specified <paramref name="operationMessageProcessor"/>.
+    /// Returns or throws <see cref="OperationCanceledException"/> when the WebSocket connection is closed.
+    /// </summary>
+    public virtual async Task ExecuteAsync(IOperationMessageProcessor operationMessageProcessor)
+    {
+        if (operationMessageProcessor == null)
+            throw new ArgumentNullException(nameof(operationMessageProcessor));
+        if (Interlocked.Exchange(ref _executed, 1) == 1)
+            throw new InvalidOperationException($"{nameof(ExecuteAsync)} may only be called once per instance.");
+        try
+        {
+            await operationMessageProcessor.InitializeConnectionAsync();
+            // set up a buffer in case a message is longer than one block
+            var receiveStream = new MemoryStream();
+            // set up a 16KB data block
+            byte[] buffer = new byte[16384];
+            // prep a Memory instance pointing to the block
+#if NETSTANDARD2_0
+            var bufferMemory = new ArraySegment<byte>(buffer);
+#else
+            var bufferMemory = new Memory<byte>(buffer);
+#endif
+            // prep a reader stream
+            var bufferStream = new ReusableMemoryReaderStream(buffer);
+            // read messages until an exception occurs, the cancellation token is signaled, or a 'close' message is received
+            while (true)
+            {
+                var result = await _webSocket.ReceiveAsync(bufferMemory, RequestAborted);
+                if (result.MessageType == WebSocketMessageType.Close)
+                {
+                    // prevent any more messages from being queued
+                    operationMessageProcessor.Dispose();
+                    // send a close request if none was sent yet
+                    if (!_outputClosed.Task.IsCompleted)
+                    {
+                        // queue the closure
+                        _ = CloseAsync();
+                        // wait until the close has been sent
+                        var completedTask = Task.WhenAny(
```

Comment:
You do not use `completedTask`. Let's throw `TimeoutException` here.

Reply:
It certainly should await the result.  But why return a timeout exception?  On this end, the closure has taken place normally.  The only reason the delay exists is to facilitate a graceful shutdown, but a non-graceful shutdown does not indicate a problem on this end.

Reply:
I added tests in my project for this scenario, which will be copied over when we copy the rest of the tests.  (Ugh; it took like an hour or more...)

Reply:
https://github.com/Shane32/GraphQL.AspNetCore3/pull/35

Reply:
I don’t think we should add OnNonGracefulShutdown at all. I have seen implementations that never perform a graceful shutdown. But if it is added, then we should determine more precisely what situations it is triggered in. 

1. Client disconnects with no warning
2. Client disconnects after server requests closure, but without client sending close message
3. Client sends close request and immediately disconnects without waiting for server close message 
4. Server sends close message, and there is a timeout waiting for client disconnect message to arrive

`OnNonGracefulShutdown(bool receivedCloseMessage, bool sentCloseMessage)`?

Reply:
Please confirm:

```cs
public virtual Task OnNonGracefulShutdownAsync(bool receivedCloseMessage, bool sentCloseMessage)
    => Task.CompletedTask;
```

Reply:
OK if this handles all cases. I did not read WebSocket protocol spec yet.

Reply:
Done.  Note that this relates to WebSockets in general, and is not specific to GraphQL over WebSockets.

Reply:
I understand it. Thanks.